File: impls/tsbun/step6_file.ts
import { readStr } from "./reader.ts";
import { pr_str_antlr } from "./printer.ts";
import { Env } from "./env.ts";
import { FunctionType, VectorType, type TispType, Node, ListType, isSeq, NumberType, True, False, equals, Nil, StringType } from "./types.ts";
import { AstNode, AtomNode } from "./AST.ts";
import { first } from "lodash/fp";
import { ns } from "./core.ts";
import { parseArgs } from "util";

const eval_ast = (ast: TispType, env: Env): TispType => {
  switch (ast.type) {
    case Node.Program:
      return ast.elements.map(node => eval_mal(node, env)).pop();
    case Node.Symbol:
      const value = env.get(ast.value);
      if (!value) {
        throw new Error(`${ast.value} not found `)
      }
      return value;
    case Node.List:
      return new ListType(ast.elements.map(node => eval_mal(node, env)));
    case Node.Vector:
      return new VectorType(ast.elements.map(node => eval_mal(node, env)));
    default:
      return ast;
  }
}

const eval_mal = (ast: TispType, env: Env): any => {
  loop:
  while (true) {
    if (ast.type !== Node.List) {
      return eval_ast(ast, env);
    }
    if (ast.elements.length < 1) {
      return ast;
    }
    // Check if special form
    const [firstParam, ...rest] = ast.elements
    switch (firstParam.type) {
      case Node.Symbol:
        switch (firstParam.value) {
          case 'def!': {
            const [key, value] = rest;
            if (key.type !== Node.Symbol) {
              throw new Error(`Expected ident but got ${key}`);
            }
            const evalValue = eval_mal(value, env);
            env.set(key.value, evalValue);
            return evalValue;
          }
          case 'let*': {
            const [bindings, body] = rest;
            if (!isSeq(bindings)) {
              throw new Error(`Expected vector but got ${bindings}`);
            }
            env = new Env(env);
            for (let i = 0; i < bindings.elements.length; i += 2) {
              if (bindings.elements[i].type !== Node.Symbol) {
                throw new Error(`Expected ident but got ${bindings.elements[i]}`);
              }
              env.set(bindings.elements[i].value, eval_mal(bindings.elements[i + 1], env));
            }
            ast = eval_mal(body, env);
            continue loop;
          }
          case 'do': {
            if (rest.length < 1) {
              return Nil;
            }
            const lastElement = rest.pop()!;
            rest.map(node => eval_mal(node, env));
            ast = lastElement;
            continue loop;
          }
          case 'if': {
            const [condition, trueBranch, falseBranch] = rest
            const evalCondition = eval_mal(condition, env);
            if (!equals(evalCondition, False) && !equals(evalCondition, Nil)) {
              ast = trueBranch;
            } else {
              if (!falseBranch) {
                ast = Nil;
              } else {
                ast = falseBranch;
              }
            }
            continue loop;
          }
          case 'fn*': {
            const [params, body] = rest;
            if (!isSeq(params)) {
              throw new Error(`Expected vector but got ${params}`);
            }
            return FunctionType.fromAst(eval_mal, env, params.elements.map(a => a.toString()), body);
          }
        }
    }
    // ...


    // otherwise Evaluate the function
    const result = eval_ast(ast, env);
    if (!isSeq(result)) {
      throw new Error(`Not a function ${result}`);
    }
    const [fn, ...args] = result.elements
    if (fn.type !== Node.Function) {
      throw new Error(`Not a function ${fn}`);
    }
    if (fn.body) {
      ast = fn.body;
      env = fn.newEnv(args);
      continue loop;
    }
    return fn.func(...args);
  }
}
const READ = (str: string): any => readStr(str);
const EVAL = (ast: any, _env?: any): any => eval_mal(ast, _env);
const PRINT = (exp: any): string => pr_str_antlr(exp)
const rep = (str: string): string => PRINT(EVAL(READ(str), rootEnv));

const repl = async () => {
  const prompt = "user> ";
  process.stdout.write(prompt);
  for await (const line of console) {
    if (line.length === 0) {
      break;
    }
    try {
      process.stdout.write(rep(line));
      process.stdout.write("\n");
    } catch (e) {
      const err: Error = e;
      process.stdout.write(err.message);
      process.stdout.write("\n");
    }
    process.stdout.write(prompt);
  }
}

const rootEnv = new Env();
Object.entries(ns).map(([key, value]) => {
  rootEnv.set(key, value);
})
rootEnv.set("eval", FunctionType.fromBootstrap((ast: TispType) => {
  if (!ast) {
    throw new Error("Cannot eval nil");
  }
  const res = eval_mal(ast, rootEnv);

  return res;
}));

rep('(def! not (fn* (a) (if a false true)))')
// load-file
rep(`(def! load-file (fn* (f) (eval (read-string (str "(do " (slurp f) "\nnil)")))))`)

const { positionals } = parseArgs({
  args: Bun.argv,
  allowPositionals: true,
})
const [_, __, ...args] = positionals;

if (args.length > 0) {
  // Read the file
  const [filename, ...rest] = args;
  rootEnv.set("*ARGV*", new ListType(rest.map(a => new StringType(a))));
  rep(`(load-file "${filename}")`);
} else {
  rootEnv.set("*ARGV*", new ListType([]));

  // Add eval to rootEnv
  repl();


}

File: impls/elixir2/lib/mix/tasks/step6_file.ex

defmodule Mix.Tasks.Step6File do
  import Mal.Types
  def run(args) do
    # Repl loop
    env = Mal.Env.new()

    bootstrap_env(args, env)
    loop(env)
  end
  defp load_file(file, env) do
    read_eval_print("(load-file \"#{file}\")", env)
    exit(:normal)
  end
  defp bootstrap_env(args, env) do
    Mal.Env.merge(env, Mal.Core.ns())
    Mal.Env.set(env, "eval", %Mal.Function{value: fn ([ast]) -> eval(ast, env) end})
    read_eval_print("""
      (def! not (fn* (a)
        (if a false true)))
    """, env)
    read_eval_print("""
      (def! load-file (fn* (f)
        (eval (read-string
          (str
            "(do "
            (slurp f)
            "\nnil)")))))
    """, env)

    case args do
      [file | args] ->
        Mal.Env.set(env, "*ARGV*", {:list, args})
        load_file(file, env)
      _ ->
        Mal.Env.set(env, "*ARGV*", {:list, []})
    end
  end

  defp loop(env) do
    IO.write(:stdio, "user> ")
    IO.read(:stdio, :line)
      |> (&read_eval_print(&1, env)).()
      |> IO.puts

    loop(env)
  end

  defp read(input) do
    Mal.Reader.read_str(input)
  end

  defp eval_ast({:symbol, ast}, env) do
    case Mal.Env.get(env, ast) do
      {:ok, val} -> val
      :not_found -> throw({:error , "Symbol #{ast} not found"})
    end
  end
  defp eval_ast({:list, ast}, env) do
    {:list, Enum.map(ast, &eval(&1, env))}
  end
  defp eval_ast({:vector, ast}, env), do: {:vector, Enum.map(ast, &eval(&1, env))}
  defp eval_ast({:map, ast}, env) do
    map = for {key, value} <- ast, into: %{} do
      {key, eval(value, env)}
    end
    {:map, map}
  end
  defp eval_ast(ast, _env), do: ast

  defp eval({:list, []} = empty_list, _env), do: empty_list

  defp eval({:list, ast}, env), do: eval_list(ast, env)

  defp eval(input, env) do
    eval_ast(input, env)
  end

  # Builtin Special Funcs

  # Define a symbol in the current environment
  defp eval_list([{:symbol, "def!"}, {:symbol, key}, value], env) do
    evaled = eval(value, env)
    Mal.Env.set(env, key, evaled)
    evaled
  end


  # Define a new environment with the current environment as parent
  # And evaluate the body in the new environment
  defp eval_list([{:symbol, "let*"}, {_, bindings}, body], env) do
    #_zipped = Enum.unzip(bindings)
    new_env = Mal.Env.new(env)

    eval_bindings(bindings, new_env)
    eval(body, new_env)
  end

  # Evaluate the list of expressions in order, and return the last one
  defp eval_list([{:symbol, "do"}| ast], env) do
    ast
    |> List.delete_at(-1)
    |> list
    |> eval_ast(env)

    eval(List.last(ast), env)
  end

  # Evaluate the condition, if true, evaluate the true_case, else evaluate the false_case
  defp eval_list([{:symbol, "if"}, condition, true_case | false_case], env) do
    case eval(condition, env) do
      n when n == nil or n == false -> case false_case do
        [] -> nil
        [body] -> eval(body, env)
      end
      _ -> eval(true_case, env)
    end
  end

  # Define a function in the current environment
  defp eval_list([{:symbol, "fn*"}, {type, params}, body], env) when type in [:list, :vector] do
    closure = fn (args) ->
      new_env = Mal.Env.new(env, params, args)
      eval(body, new_env)
    end

    %Mal.Function{value: closure}
  end



  defp eval_list(ast, env) do
    {:list, [func | args]} = eval_ast({:list, ast}, env)
    func.value.(args)
  end

  defp eval_bindings([], _), do: []
  defp eval_bindings([{:symbol, key}, value | rest], env) do
    Mal.Env.set(env, key, eval(value, env))
    eval_bindings(rest, env)
  end


  defp print(input) do
    Mal.Printer.print_str(input)
  end

  defp read_eval_print(:eof, _), do: exit(:normal)
  defp read_eval_print(line, env) do
    # env = %{
    #   "+" => fn (a, b) -> a + b end,
    #   "-" => fn (a, b) -> a - b end,
    #   "*" => fn (a, b) -> a * b end,
    #   "/" => fn (a, b) -> a / b end,
    # }

    read(line)
      |> (&eval(&1, env)).()
      |> print
  catch
    {:error, msg} -> "Error: #{msg}"
  end

end
